pkg/credentialprovider/azure/azure_acr_helper.go

```diff
@@ -0,0 +1,339 @@
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE
+*/
+
+// Source: https://github.com/Azure/acr-docker-credential-helper/blob/a79b541f3ee761f6cc4511863ed41fb038c19464/src/docker-credential-acr/acr_login.go
+
+package azure
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+type authDirective struct {
+	service string
+	realm   string
+}
+
+type accessTokenPayload struct {
+	TenantID string `json:"tid"`
+}
+
+type acrTokenPayload struct {
+	Expiration int64  `json:"exp"`
+	TenantID   string `json:"tenant"`
+	Credential string `json:"credential"`
+}
+
+type acrAuthResponse struct {
+	RefreshToken string `json:"refresh_token"`
+}
+
+// 5 minutes buffer time to allow timeshift between local machine and AAD
+const timeShiftBuffer = 300
+const userAgentHeader = "User-Agent"
+
+const tokenUsername = "<token>"
+const tokenUsernameGUID = "00000000-0000-0000-0000-000000000000"
+
+var client = &http.Client{}
+var userAgentVersion string
+
+func getUserAgent() string {
+	if userAgentVersion == "" {
+		return "acr-docker-credential-helper"
+	}
+	return userAgentVersion
+}
+
+func (token *acrTokenPayload) isExpiredOrNear() bool {
+	return time.Now().Unix() > token.Expiration-timeShiftBuffer
+}
+
+// GetUsernamePassword get the AAD based ACR login credentials
+func GetUsernamePassword(serverAddress string, identityToken string) (user string, cred string, err error) {
+	if identityToken == "" {
+		return "", "", fmt.Errorf("Unexpected empty token. Please call 'az acr login' to generate a valid token")
+	}
+
+	var acrToken *acrTokenPayload
+	acrToken, err = parseAcrToken(identityToken)
+	if err != nil {
+		return "", "", fmt.Errorf("Bad identity token")
+	}
+	refreshToken := identityToken
+	if acrToken.isExpiredOrNear() {
+		var challenge *authDirective
+		if challenge, err = receiveChallengeFromLoginServer(serverAddress); err != nil {
+			// ignore all error when receiving the challenge
+			logrus.Infof("[Azure Login Helper] server %s didn't respond with a valid challenge, reverting to default login...\nerror: %s\n", serverAddress, err)
+			return "", "", nil
+		}
+		refreshToken, err = performTokenExchange(serverAddress, challenge, acrToken.TenantID, acrToken.Credential)
+	}
+
+	return tokenUsername, refreshToken, nil
```

Comment:
But yeah, the function in general is a bit confusing. I think the `<token>` thing has more to do with the docker/cred-helper apis rather than the ACR API.